components/script/textinput.rs

```diff
@@ -1119,28 +1120,34 @@ impl<T: ClipboardProvider> TextInput<T> {
         TextPoint { line, index }
     }
 
+    /// Set the current selection range and direction
+    ///
+    /// Both `start` and `end` are indices of code points, not bytes.
     pub fn set_selection_range(&mut self, start: u32, end: u32, direction: SelectionDirection) {
-        let mut start = UTF8Bytes(start as usize);
-        let mut end = UTF8Bytes(end as usize);
-        let text_end = self.get_content().len_utf8();
+        let content = self.get_content();
 
-        if end > text_end {
-            end = text_end;
-        }
-        if start > end {
-            start = end;
-        }
+        let byte_start = content
+            .chars()
+            .take(start as usize)
+            .fold(0, |acc, e| acc + e.len_utf8());
+        let byte_end = content
+            .chars()
+            .take(end as usize)
+            .fold(0, |acc, e| acc + e.len_utf8());
```

Comment:
This could be made more efficient by not counting the characters in the segment before `start` twice. But it's a bit fiddly because `start` may be larger than `end` and feels like premature optimization at this point.